### ISSUE TITLE
Asking for feedback on a JSON log file format

### ISSUE THREAD

**Issue by marcelm** [ewels/MultiQC#1482]
(Please let me know if this should be discussed elsewhere.)

I’ve started work on adding an option for writing a report in JSON format to Cutadapt. I added this because I want to easily be able to grab the number of trimmed reads and similar, without having to mess with regular expressions, but I’ve designed the format also with MultiQC in mind.

(@ewels, this is in part your idea because a couple of years ago, during your first presentation of MultiQC, you complained about hard to parse log files. I’ve been wanting to add some kind of machine-readable report to Cutadapt ever since.)

Now after seeing #1328, I realized it would make sense to get some feedback before releasing that Cutadapt version.

An example report is [shown in the documentation](https://cutadapt.readthedocs.io/en/latest/guide.html#json-report). The description is incomplete at the moment, but you should be able to deduce the rest. Please let me know what you think.

Some other points I wanted to raise:
- The report is currently only created if the user provides `--json=filename.cutadapt.json`, so would it actually make sense to support this as another report format in MultiQC? The "normal" report would need to be supported anyway, and at the moment the JSON file doesn’t contain extra information. And there would need to be a mechanism that ignores the regular log file if a JSON report exists.
- Would it instead be more desirable to *always* print the JSON output as part of the regular report? It could be quite compact when printed on one line without unnecessary whitespace, but it would look a bit ugly.

**Comment by ewels** [comment]
I love it! I wrote a blog post to get my thoughts down a while back: http://tallphil.co.uk/writing-good-log-files/ (a little more than a couple of years ago now 😉 ) and you're already hitting most of the points. If I'm being totally selfish from MultiQC's point of view, it would be nice to request a couple of additional features:

* Instead of specifying a filename, specify a prefix so that all summary files _always_ end in `.cutadapt.json`. This makes the files much easier to find, as MultiQC doesn't have to scrape file contents looking for strings (which is slow).
* Produce the file by default, as many people don't RTFM and maybe realise that they need this file for MultiQC until _after_ they have run trimming.
    * The downside is "polluting" the output directory with files I guess. However, I doubt that many people run cutadapt in an interactive session, so I would presume that the vast majority of people are already redirecting the stderr log to a file. My bet would be that the majority of users would be pleasantly surprised..

Having both outputs should not be an issue as long as the sample names end up identical (which they should be, as they're based on the filename pulled from the cutadapt command). We can run the JSON parsing second and it will then overwrite anything that has been picked up on the stderr logs. We can stop official support / maintenance of the stderr log and over time people will stop saving it / passing it to MultiQC.

I'm not sure that printing the JSON within the regular report (stderr) really adds anything from the point of view of MultiQC, so personally I wouldn't bother.

For the JSON output itself, I think it looks fab (love that you keep the keys with `null` values 👍🏻 ). Two minor points:

* For numbers that sum to a total, I think it's nice to have them grouped for clarity. Then people like me don't have to figure how to get them to add up properly. Specifically here I'm thinking for the filtering, so for example maybe instead:
    ```json
    "read_counts": {
        "input": 1000000,
        "output": 981620,
        "reverse_complemented": null,
        "read1_with_adapter": 233500,
        "read2_with_adapter": null,
        "filtered": 18380,
        "filter_categories": {
            "too_short": 18380,
            "too_long": null,
            "too_many_n": null,
            "too_many_expected_errors": null,
            "casava_filtered": null
        }
    },
    ``` 
* The adapter trimming sections seem to be missing a few stats that are in the existing stderr log.. Specifically `Overview of removed sequences` which is used by MultiQC for the trimming length plot. Also the proportions for all 4 bases (though MultiQC doesn't use that).
* Any reason that `on_reverse_complement` is `0` instead of `false`?

Great that you're adding this, and thanks so much for taking the time to chat about it before it goes live! 🙏🏻

**Comment by marcelm** [comment]
> I love it! I wrote a blog post to get my thoughts down a while back: http://tallphil.co.uk/writing-good-log-files/ (a little more than a couple of years ago now wink ) and you're already hitting most of the points.

Oh, I remember now! I read it back then.

> * Instead of specifying a filename, specify a prefix so that all summary files _always_ end in `.cutadapt.json`. This makes the files much easier to find, as MultiQC doesn't have to scrape file contents looking for strings (which is slow).

I have to say I’m not too comfortable with automatically adding the suffix because I think that would be unexpected behavior. For many other command-line tools, you have to give full file names. It was my intention, though, that everyone uses the `.cutadapt.json` extension. I will make this clearer in the documentation.

> * Produce the file by default, as many people don't RTFM and maybe realise that they need this file for MultiQC until _after_ they have run trimming.

You call this the "most controversial" suggestion in your blog post, and I unfortunately have to agree: I think a well-behaved tool that essentially just streams from input and writes to output shouldn’t write extra files anywhere without being told to do so explicitly. I’m not totally against it, but it would need some more convincing.

What I’m ready to do already now is to add an option `-J` (or whichever letter is still available) that would tell Cutadapt to dump the JSON log to "name of R1 output file" + `.cutadapt.json` (when output is stdout, that behavior would be disabled). Then `--json` would only be used to override the name of that file.

(And I would add a MultiQC section to the documentation that recommends always using that option.)

> The downside is "polluting" the output directory with files I guess. However, I doubt that many people run cutadapt in an interactive session, so I would presume that the vast majority of people are already redirecting the stderr log to a file. 

Because people already redirect stderr to a log, I would have thought that always embedding the JSON in there would help. You’d still need to find the proper file and then the correct line with the JSON part, but you wouldn’t have to parse any of the other lines.

> Having both outputs should not be an issue as long as the sample names end up identical (which they should be, as they're based on the filename pulled from the cutadapt command). We can run the JSON parsing second and it will then overwrite anything that has been picked up on the stderr logs. We can stop official support / maintenance of the stderr log and over time people will stop saving it / passing it to MultiQC.

Would you stop support for the stderr log even if I decide not to always write the JSON file?

> I'm not sure that printing the JSON within the regular report (stderr) really adds anything from the point of view of MultiQC, so personally I wouldn't bother.
> 
> For the JSON output itself, I think it looks fab (love that you keep the keys with `null` values 👍🏻 ). Two minor points:
> 
> * For numbers that sum to a total, I think it's nice to have them grouped for clarity. Then people like me don't have to figure how to get them to add up properly. Specifically here I'm thinking for the filtering, so for example maybe instead:

(`filter_categories` added)

Good idea, I will change that.

> * The adapter trimming sections seem to be missing a few stats that are in the existing stderr log.. Specifically `Overview of removed sequences` which is used by MultiQC for the trimming length plot. Also the proportions for all 4 bases (though MultiQC doesn't use that).

Adding the table with the lengths of the removed sequences made the JSON file explode, which is why that particular code is commented out at the moment. I need to find a way to serialize this in a nicer way because I would like the JSON to be somewhat human-readable. But I’ll add it even if I don’t figure out how to make it pretty.

> * Any reason that `on_reverse_complement` is `0` instead of `false`?

It’s the number of adapter occurrences found on the reverse complement of the read. This is actually only relevant when `--revcomp` is used, so here it should be `null`. I’ll fix that as well.

> Great that you're adding this, and thanks so much for taking the time to chat about it! 🙏🏻

Thanks for the feedback!

It’s gotten late; I’ll add Cutadapt issues for the above to-do items later this week.

**Comment by ewels** [comment]
> I’m not totally against it, but it would need some more convincing.

Yeah I figured that you might not like this idea! 😉  My most convincing argument is that the author of HiSAT2 added a `--new-summary` command line option after a similar conversation. It is my suspicion that 99% of users are not aware of it and do not use it. As such, MultiQC still has to maintain support for both formats and the benefits of the nicer output are somewhat lost.

This doesn't mean that having the JSON isn't still an improvement - it definitely is. It's just that I am a cynical sod and I doubt that many people will use it as most have the commands that they are used to and wouldn't go looking for this one.

Quite a few tools do produce summary files by default - in my opinion, it's increasingly valuable to generate these machine-readable summary files as genomics datasets get bigger (eg. single cell stuff), and I suspect that's partly why.  But, yes it's probably not best unix practice.

> What I’m ready to do already now is to add an option `-J` (or whichever letter is still available) that would tell Cutadapt to dump the JSON log to "name of R1 output file" + `.cutadapt.json` (when output is stdout, that behavior would be disabled). Then `--json` would only be used to override the name of that file.

Definitely a good compromise for the filename suffix point 👍🏻 Especially if you can also do `--json` as well as `-J` and just not give a filename.

> Because people already redirect stderr to a log, I would have thought that always embedding the JSON in there would help.

Yup - if you're not always creating the JSON by default for all users, then it could be nice to print to stdout 👍🏻 Then we get the benefits of parsing the nice JSON there too. It's just not quite as clean / easy to access.

> Would you stop support for the stderr log even if I decide not to always write the JSON file?

I doubt I would _remove_ support for the stderr log any time soon. People run terrifyingly old versions of cutadapt sometimes! You can see from a couple of other issues I linked here earlier that the JSON format should solve a few issues that people have been hitting in MultiQC though, so I will stop bothering trying to fix those and just tell people to use JSON. Then maybe remove the stderr stuff a bit further down the line once everyone is using it.

**Comment by ewels** [comment]
Out of curiosity - what's the reason to make it `--json=filename.cutadapt.json` instead of `--json filename.cutadapt.json`?

One thought I had last night was that MultiQC could print a warning in the log if stdout logs are parsed from a recent version of cutadapt and not JSON. It's a bit of a long shot, but that might also help people adopt a new flag..

**Comment by marcelm** [comment]
I’ll spend more time on this later, but just to answer the last point: This is a how it’s suggested for GNU tools, see https://www.gnu.org/software/libc/manual/html_node/Argument-Syntax.html . I prefer that style because it is clear immediately when you see a command that this is an option and its argument. When you see `--json filename`, that could be an option without argument plus a positional argument. `rsync`, which is my role model for a tool that has many command-line arguments, uses that style. Also, `optparse` prints it that way in the autogenerated help. That got unfortunately lost when I switched to `argparse`.

Note that you can write it both ways. `argparse` accepts it either way.

**Comment by marcelm** [comment]
It’s been a while, but I’ve spent some time on this over the last days. A suprising amount of refactoring was necessary to make this easy to generate. I even implemented a custom JSON serializer so that the file would look nice (not too many linebreaks). The description of the file format is available at https://cutadapt.readthedocs.io/en/latest/reference.html#json-report-format (note the URL has changed), including an example, that I’ll copy the example verbatim below.

I think I took into account all of your comments, but I just noticed hat MultiQC is actually using the "expected number of trimmed sequences" column in the histogram, so I’ll add that as well.

Just FYI, you don’t need to review this again.

(I haven’t added the `-J` option; possibly that will have to wait for the next release.)
```json
{
  "tag": "Cutadapt report",
  "schema_version": [0, 1],
  "cutadapt_version": "3.5",
  "python_version": "3.8.10",
  "command_line_arguments": [
    "--json=out.cutadapt.json", "-m", "20", "-a", "ACGTAC", "-q", "20",
    "--discard-trimmed", "-o", "out.fastq.gz", "reads.fastq"],
  "cores": 1,
  "input": {
    "path1": "reads.fastq",
    "path2": null,
    "paired": false,
    "interleaved": null
  },
  "read_counts": {
    "input": 100000,
    "filtered": {
      "too_short": 251,
      "too_long": null,
      "too_many_n": null,
      "too_many_expected_errors": null,
      "casava_filtered": null,
      "discard_trimmed": 2061,
      "discard_untrimmed": null
    },
    "output": 97688,
    "reverse_complemented": null,
    "read1_with_adapter": 2254,
    "read2_with_adapter": null
  },
  "basepair_counts": {
    "input": 10100000,
    "input_read1": 10100000,
    "input_read2": null,
    "quality_trimmed": 842048,
    "quality_trimmed_read1": 842048,
    "quality_trimmed_read2": null,
    "output": 9038081,
    "output_read1": 9038081,
    "output_read2": null
  },
  "adapters_read1": [
    {
      "name": "1",
      "type": "regular_three_prime",
      "specification": "ACGTAC",
      "total_matches": 2254,
      "on_reverse_complement": null,
      "five_prime_statistics": null,
      "three_prime_statistics": {
        "error_rate": 0.1,
        "error_lengths": [6],
        "matches": 2254,
        "adjacent_bases": {
          "A": 473,
          "C": 1240,
          "G": 328,
          "T": 207,
          "": 6
        },
        "dominant_adjacent_base": null,
        "trimmed_lengths": [
          {"len": 3, "counts": [1220]},
          {"len": 4, "counts": [319]},
          {"len": 5, "counts": [30]},
          {"len": 6, "counts": [4]},
          {"len": 7, "counts": [5]},
          {"len": 8, "counts": [7]},
          {"len": 9, "counts": [4]},
          {"len": 10, "counts": [7]},
          {"len": 11, "counts": [7]},
          {"len": 12, "counts": [6]},
          {"len": 13, "counts": [8, 2]},
          {"len": 14, "counts": [1, 1]},
          {"len": 15, "counts": [2, 0]},
          {"len": 16, "counts": [3, 1]},
        ]
      }
    }
  ],
  "adapters_read2": null
}
```

**Comment by NanoCoreUSA** [comment]
Was this ever implemented in the current version of MultiQC? If not, is it still planned to be incorporated?

I've been outputting the JSON file that cutadapt produces (--json=filename.cutadapt.json) but I still have to rely on old 'txt' file to be parsed by MultiQC.

**Comment by ewels** [comment]
Not yet, no. Might be worth creating a new issue about this if you could. I'll need some example test data (don't have any JSON files [here](https://github.com/ewels/MultiQC_TestData/tree/master/data/modules/cutadapt) yet).
